c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -2837,14 +2837,38 @@ void local_raw_delete(void* ptr);
 
 class NativeCachingAllocator : public CUDAAllocator {
  private:
-  std::mutex mutex;
+  static constexpr size_t kNumMutexShard = 1031;
```

Comment:
Any study on why this value? Any minimum level of concurrency we expect to smoothly support?

Reply:
Same question, wondering how we come up with this magic number?

Reply:
I think it's possible to reduce this number without affecting lock contention, it's based on e2e tests so not super accurate and other workloads might be different

Reply:
yeah, wondering if we reduce it to 63, how the perf looks like? I feel 1031 may be an overkill?

Reply:
Any concern about this array not fitting in a cache line or something similar that would push us to make it smaller?
If not, then I guess the justification is that we need whatever is "big enough" and, empirically, this is enough. But users can increase it if they see contention issues still happening.

Reply:
ok let me kept it in 63?